src/x86_64/schedule.c

```diff
@@ -0,0 +1,182 @@
+#include <kernel.h>
+#include <tfs.h> // needed for unix.h
+#include <unix.h> // some deps
+#include <apic.h>
+
+
+/* Try to keep these within the confines of the runloop lock so we
+   don't create too much of a mess. */
+//#define SCHED_DEBUG
+#ifdef SCHED_DEBUG
+#define sched_debug(x, ...) do {log_printf("SCHED", "[%2d] " x, ci->id, ##__VA_ARGS__);} while(0)
+#else
+#define sched_debug(x, ...) (void)ci
+#endif
+
+// currently defined in x86_64.h
+static char *state_strings_backing[] = {
+    "not present",
+    "idle",
+    "kernel",
+    "interrupt",
+    "user",         
+};
+
+char **state_strings = state_strings_backing;
+static int wakeup_vector;
+int shutdown_vector;
+
+queue runqueue;                 /* kernel space from ?*/
+queue bhqueue;                  /* kernel from interrupt */
+queue thread_queue;             /* kernel to user */
+timerheap runloop_timers;
+u64 idle_cpu_mask;              /* xxx - limited to 64 aps. consider merging with bitmask */
+timestamp last_timer_update;
+
+static timestamp runloop_timer_min;
+static timestamp runloop_timer_max;
+
+static u64 kernel_lock;
+
+void kern_lock()
+{
+    cpuinfo ci = current_cpu();
+    spin_lock(&kernel_lock);
+    ci->have_kernel_lock = true;
+}
+
+boolean kern_try_lock()
+{
+    cpuinfo ci = current_cpu();
+    if (ci->have_kernel_lock)
+        return true;
+    if (!spin_try(&kernel_lock))
+        return false;
+    ci->have_kernel_lock = true;
+    return true;
+}
+
+void kern_unlock()
+{
+    cpuinfo ci = current_cpu();
+    assert(ci->have_kernel_lock);
+    ci->have_kernel_lock = false;
+    spin_unlock(&kernel_lock);
+}
+
+static void run_thunk(thunk t, int cpustate)
+{
+    cpuinfo ci = current_cpu();
+    sched_debug(" run: %F state: %s\n", t, state_strings[cpustate]);
+    // as we are walking by, if there is work to be done and an idle cpu,
+    // get it to wake up and examine the queue
+    if (idle_cpu_mask &&
+        ((queue_length(bhqueue) > 0) ||
+         (queue_length(runqueue) > 0) ||
+         (queue_length(thread_queue) > 0))) {
+        // unfortunately, if idle cpu mask is zero (which can happen since this
+        // is racy), the result is the previous value ... so asm here
```

Comment:
Not clear why it would be more than just a nop...if the target cpu has something already scheduled at kernel level, it really shouldn't be invoking runloop anyway. That would be something like a preemptible kernel, which I don't see happening.

That said, this idle cpu selection could pretty easily be made into a cas loop to avoid a spurious wakeup...I think...